src/scene/materials/shader-graph-builder.js

```diff
@@ -0,0 +1,350 @@
+     */
+    addConst: function (type, userId, value) {
+        var name = userId ? 'const_' + type + '_' + userId : '';
+        if (userId && this._addedNamedInputs[name]) {
+            console.error('pc.ShaderGraphBuilder#addConst: userId ' + userId + ' already used!');
+            return null;
+        }
+        if (!value) {
+            console.error('pc.ShaderGraphBuilder#addConst:' + (userId ? '' : type) + ' constant ' + (userId || '') + ' has no value!');
+            return null;
+        }
+
+        var ioPort = this._graph.addConstant(type, userId, value);
+
+        if (userId) this._addedNamedInputs[name] = ioPort;
+
+        return ioPort;
+    },
+    _validateArg: function (arg, err) {
+        if (arg.type) {
+            // convert ioPort to port node
+            arg = { node: arg };
+        }
+        if ((arg.node && arg.node.type)) {
+            // port node
+            if (!this._addedNamedInputs[arg.node.name] && !arg.node.name.startsWith('const_') ) {
+                console.error(err + ": invalid input param: " + arg.node.name);
+                return false;
+            }
+        } else {
+            if (typeof(arg) === 'number') {
+                arg = { node: arg, port: 'ret' };
+            }
+
+            if (!arg.port) {
+                arg = { node: arg.node, port: 'ret' };
+            }
+
+            var core = this._addedNodes[arg.node];
+            if (!core) {
+                console.error(err + ": invalid node index: " + arg.node);
+                return false;
+            }
+            var port = core.getIoPortByName(arg.port);
+            if (!port) {
+                console.error(err + ": invalid output port: " + core.name + "." + arg.port);
+                return false;
+            }
+        }
+        return true;
+    },
+
+    _getArgTypes: function (args, err) {
+        var argTypes = [];
+        var argIndex;
+        for (argIndex = 0; argIndex < args.length; argIndex++) {
+            var arg = args[argIndex];
+
+            if (arg.type) {
+                // direct ioPort - no swizzle
+                arg = { node: arg }; // node is a port node
+            }
+            if (typeof(arg) === 'number') {
+                // node id - no swizzle - set ret port
+                arg = { node: arg, port: 'ret' };
+            }
+
+            var unswizzledType;
+            if (arg.node && arg.node.type) {
+                // port node - probably with swizzle
+                unswizzledType = arg.node.type;
+            } else {
+                if (!arg.port) {
+                    arg = { node: arg.node, port: 'ret', swizzle: arg.swizzle };
+                }
+
+                // node and port - maybe with swizzle
+                var core = this._addedNodes[arg.node];
+                if (!core) {
+                    console.error(err + ": invalid node index: " + arg.node);
+                    return false;

```

Comment:
the function returns an array .. so for type consistency, this should return null instead of false.

Reply:
I'll change it

Reply:
done